### PR TITLE
Refactor IoC.

### PR DESCRIPTION
**Pull request by PJB3005**
Is now attribute driven and more compile time enforced (with interfaces). IoC can now resolve types (and create them with a helper to reduce boilerplate).

Should allow for better modularization and stuff later down the line.

### CODE REVIEW

**Comment by PJB3005** [comment]
@ComicIronic Should I make `[IoCTarget]` required or optional? Right now it's optional.

File: SS14.Client.Services/Configuration/ConfigurationManager.cs [hunk 1]
```diff
@@ -2,9 +2,11 @@
 using SFML.Window;
 using System.IO;
 using System.Xml.Serialization;
+using SS14.Shared.IoC;
 
 namespace SS14.Client.Services.Configuration
 {
+    [IoCTargetAttribute]
```
Comment (ComicIronic) [inline SS14.Client.Services/Configuration/ConfigurationManager.cs]: `IoCTarget`?

**Comment by ComicIronic** [comment]
Optional for what, to actually get the type? It should definitely be mandatory for that, I don't want to shim a test implementation and it gets used in the game by accident.

**Comment by PJB3005** [comment]
@ComicIronic alright then.

File: SS14.UnitTesting/SS14.Shared/IoC/IoCManager_Test.cs [hunk 2]
```diff
@@ -38,6 +38,30 @@ public void IoCTestAttributes()
...
+                {
+                    did1 = true;
+                }
+                else if (!did2 && type == typeof(IoCTestPriorities2))
+                {
+                    did2 = true;
+                }
+                else
+                {
+                    Assert.Fail("IoCManager returned too much types.");
```
Comment (ComicIronic) [inline SS14.UnitTesting/SS14.Shared/IoC/IoCManager_Test.cs]: many* , minor point.

**Comment by psykzz** [comment]
Has the usage changed with this PR?

If so, can you do a small example for clarity?

**Comment by PJB3005** [comment]
@psykzz yes.

To allow IoC to resolve an interface, the interface needs to inherit `IIoCInterface`:
```cs
public interface IFoo : IIoCInterface { ... }
```

To allow IoC to find an implementation, said implementation needs the `[IoCTarget]` attribute:
```cs
[IoCTarget]
public class Bar : IFoo { ... }
```

**Comment by psykzz** [comment]
Do you still use `IoCManager.Resolve<T>()` ?

**Comment by PJB3005** [comment]
Yes.

There's also `ResolveType<T>()` which returns a type instead of an instance, `NewType<T>()` which returns a *new* instance (no services), and `ResolveEnumerable<T>()` which returns an `IEnumerable<Type>` over all implementors.

File: SS14.UnitTesting/SS14.Shared/IoC/IoCManager_Test.cs (hunk 2 above)
Comment (ZoldorfTheWizard) [reply SS14.UnitTesting/SS14.Shared/IoC/IoCManager_Test.cs]: code good too much types :P

**Pull request merged**